packages/flutter_tools/lib/src/resident_runner.dart

```diff
@@ -1502,6 +1502,68 @@ abstract class ResidentRunner extends ResidentHandlers {
     appFinished();
   }
 
+  @protected
+  Future<void> performPreHotRestart() async {
+    final Timer slowCallbackTimer = Timer(const Duration(seconds: 5), () async {
+      globals.printError('preHotRestartCallbacks are taking longer than expected...');
+    });
+    final List<Future<void>> preHotRestartFutures = <Future<void>>[];
+    for (final FlutterDevice device in flutterDevices) {
+      final FlutterVmService? vmService = device.vmService;
+      if (vmService == null) {
+        continue;
+      }
+      final List<FlutterView> views = await vmService.getFlutterViews();
+      for (final FlutterView view in views) {
+        final String? isolateId = view.uiIsolate?.id;
+        if (isolateId == null) {
+          continue;
+        }
+
+        // Make sure that we are able to run the preHotRestart callbacks,
+        // by unpausing the isolate and removing breakpoints, otherwise we will be
+        // infinitely waiting for the callbacks to run.
+        await unpauseIsolateAndRemoveBreakpoints(vmService, isolateId);
```

Comment:
will this reset a user's breakpoints?

Reply:
we'd need to make sure we re-enabled the breakpoints we disabled because otherwise it's basically dataloss (of the user's carefully crafted breakpoints).

Reply:
@Hixie  This doesn't remove the breakpoints, as per this comment https://github.com/flutter/flutter/blob/532a84cd1554186c7d5ee5afef889273a7b80393/packages/flutter_tools/lib/src/run_hot.dart#L625

The function is the same piece of code from the regular hot restart code, but extracted into its own so that it can be used in the pre hot restart callbacks logic as well.

Reply:
if it doesn't remove the breakpoints may i suggest a different name than "unpauseIsolateAndRemoveBreakpoints" :-)

Reply:
also maybe change the comment not to say that it is "unpausing the isolate and removing breakpoints".

Reply:
I'm not really sure how to name it. It does remove the breakpoints in the isolate. It just happens that hot restart spawns a new isolate, and then the editor restores the breakpoints. Maybe we can make this clearer with a comment in the 2 places it is called and keep the removeBreakpoints name.
What do you think?

Reply:
I added some explanation here 7ab56ffe8aa0ed08721620f8359cbc158a13ae96